docs/modules/SDDs/pages/0033-commodore-component-instance-versioning.adoc

```diff
@@ -0,0 +1,91 @@
+
+If the instance version specifies a different URL from the base component, more work is needed: in this case, we clone the changed repository for the instance and create a worktree from that.
+
+This means that if two instances both override the URL to the same value, this repository is cloned twice, which can impact performance.
+However, since this only affects instances which override the repository URL, the impact is tolerable.
+
+=== Risks and Drawbacks
+
+* Since the JSonnet dependencies and libraries are all provided from a single component version, there is a risk for some combinations of versions not compiling properly due to version compatibility issues.
+** Even if a component's dependencies are sufficiently backwards-compatible at the moment, it's possible for future changes to break compatibility.
+This might force a component author to remove multi-version support in higher component versions, causing a catalog to stop compiling if either component instance is updated.
+* Overriding the repository URL in a component instance always causes a new repository to be cloned - even if the same URL is used in multiple instances.
+
+== Alternatives
+
+=== Automatically detecting multi-version support
+
+It could be possible to automatically determine whether a component supports multi-versioning, by way of checking whether the `${_base_directory}` parameter is used in the component's compile configuration.
+However, automatically declaring components to be "safe for multi-versioning" carries some risk - even if a component is configured accordingly, it is not guaranteed that its provided JSonnet libraries and/or its JSonnet dependencies work if used across versions.
+Therefore, we deem it safer to let component authors decide if multi-versioning is safe to use for their component.
+
+=== Supporting different versions of JSonnet dependencies per instance
+
+We could possibly achieve support for separate Jsonnet dependency versions per instance by reworking how we interface with Kapitan's `compile()` function.
+This would require more extensive reworking of Commodore.
+
+Making this change has no direct impact on the implementation of the multi-versioning feature, so support can still be added at a future time if required.
+
+=== Supporting different versions of a component's provided JSonnet libraries per instance
+
+It might be possible to allow multiple versions of a component's JSonnet libraries to be made available to other components.
+Other components would have to refer to the libraries using a specific instance's name.
+This would change how components refer to libraries, and would cause components to be somewhat aware of the context in which they are used in order to refer to the correct instances.
+Since components should be modular and reusable, introducing this sort of coupling goes against the design principles of Project Syn.
```

Comment:
Let me know if I'm understanding this correctly. To me this approach just sounded ugly and like a terrible idea.